api/v1alpha1/accessrequest_types.go

```diff
@@ -20,28 +20,77 @@ import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
-// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
-// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+// Status defines the different stages a given access request can be
+// at a given time.
+// +kubebuilder:validation:Enum=requested;granted;expired;refused
+type Status string
+
+const (
+	// RequestedStatus is the stage that defines the access request as pending
+	RequestedStatus Status = "requested"
+
+	// GrantedStatus is the stage that defines the access request as granted
+	GrantedStatus Status = "granted"
+
+	// ExpiredStatus is the stage that defines the access request as expired
+	ExpiredStatus Status = "expired"
+
+	// RefusedStatus is the stage that defines the access request as refused
+	RefusedStatus Status = "refused"
+)
 
 // AccessRequestSpec defines the desired state of AccessRequest
 type AccessRequestSpec struct {
-	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
-	// Important: Run "make" to regenerate code after modifying this file
+	// Duration defines the ammount of type that the elevated access
+	// will be granted once approved
+	Duration metav1.Duration `json:"duration"`
+	// TargetRoleName defines the role name the user will be assigned
+	// to once the access is approved
+	TargetRoleName string `json:"targetRoleName"`
+	// Application defines the Argo CD Application to assign the elevated
+	// permission
+	Application TargetApplication `json:"application"`
+	// Subjects defines the list of subjects for this access request
+	Subjects []Subject `json:"subjects"`
+}
+
+// TargetRoleName defines the role name the user will be assigned
+// to once the access is approved
+type TargetApplication struct {
+	// Name refers to the Argo CD Application name
+	Name string `json:"name"`
+	// Namespace refers to the namespace where the Argo CD Application lives
+	Namespace string `json:"namespace"`
+}
 
-	// Foo is an example field of AccessRequest. Edit accessrequest_types.go to remove/update
-	Foo string `json:"foo,omitempty"`
+// Subject defines the user details to get elevated permissions assigned
+type Subject struct {
+	// GroupClaim refers to the necessary OIDC 'groups' claim the user needs to
+	// have associated with their ID Token
+	GroupsClaim string `json:"groupsClaim"`
```

Comment:
If I recall correctly, the Subject is the user we should grant elevated permission to. I think from argo's perspective, username/email makes sense.
```suggestion
	// Username refers to the entity requesting the elevated permission
	Username string `json:"username"`
```

Reply:
The problem is that the username is present depending on the configuration. Even if we filter by the username in our config, in the code we are basically iterating over the GroupsClaim. That is why I thought about naming the field closer to what it is. WDYT?

Reply:
I would abstract this and not expose that logic in the API. My idea would be to only send "who" is the request for to the server, and let the server/plugin decide if and "how" that "who" has permissions. If the server decides to use the group claims, the server should have a way to fetch the claims based on the userId IMO and the server should be configured to know which claims to use. I dont think it should be part of the request.

Reply:
reverted the change.